File: chap14_respiration/O2_transport.m

% Keener and Sneyd. Mathematical Physiology, 3rd edition, Chapter 14,
% Figure 1.4.

% The graph should be computed in concentrations and then converted to
% pressures using the solubility. Be careful with the units. It is not a good idea
% to work in units of Molar. The values for sigma are much too
% small, and the numerics start going a little weird. So   all
% units are converted to microMolar. The code plays much more nicely.

function O2_transport

clearvars
close all
clc

set(0,                           ...
   'defaultaxesfontsize', 20,   ...
   'defaultaxeslinewidth', 1.0, ...
   'defaultlinelinewidth', 2.0);

global p

%% The figure in the text
p.sigmaO = 1.4;      % Units of uMolar/mm Hg
p.KO = 30*p.sigmaO;
p.PO2 = 104*p.sigmaO;   % convert pressure to conc (units of uMolar)
p.THb = 2000;           % 2 mM
L = 100;
xspan = linspace(0,L,2000);
init = 40*p.sigmaO;
[X,O] = ode15s(@rhs,xspan,init);

figure(1)
    plot(X,O/p.sigmaO,'r')          % plot the pressure, not the conc
    xlabel('dimensionless distance')
    ylabel('[O_2]')
    hold on

%% This next part is for one of the Exercises.

% Find the solution without the equilibrium approximation. When solving
% this way, be careful with  initial conditions, and be sure 
% to start on the critical manifold. Otherwise, the solution takes some
% time to catch up and there is an offset in the solution. 

timescale = 10;
p.k1 = timescale;
p.km1 = p.KO^4*p.k1;
Oinit = 40*p.sigmaO;
Yinit = p.THb*Oinit^4/(p.KO^4 + Oinit^4);
init = [Oinit, Yinit];
opts = odeset('RelTol',1e-7,'AbsTol',1e-7,'MaxStep',0.01);
[X,sol] = ode15s(@twodrhs,xspan,init,opts);
O = sol(:,1); Y = sol(:,2);

plot(X,O/p.sigmaO,'--b')          % plot the pressure, not the conc
xlabel('dimensionless distance')
ylabel('[O_2]')
hold off

end % of main

%% 
function out = rhs(x,O)
    global p
    fprime = 4*O^3*p.KO^4/( (p.KO^4 + O^4)^2 );
    out = (p.PO2 - O)/((1+4*p.THb*fprime));
end

%%
function out = twodrhs(x,dum)
    global p
    O = dum(1); Y = dum(2);
    X = p.THb - Y;
    out(1) = p.PO2 - O + 4*p.km1*Y - 4*p.k1*X.*O.^4;
    out(2) = p.k1*X*O.^4 - p.km1*Y;
    out = out';
end

File: chap14_respiration/O2_transport.py

#  Keener and Sneyd. Mathematical Physiology, 3rd edition, Chapter 14,
#  Figure 1.4.

# Written by James Keener and James Sneyd.

#  The graph should be computed in concentrations and then converted to
#  pressures using the solubility. Be careful with the units. It is not a good idea
#  to work in units of Molar. The values for sigma are much too
#  small, and the numerics start going a little weird. So   all
#  units are converted to microMolar. The code plays much more nicely.

import numpy as np
import matplotlib.pyplot as plt
from scipy.integrate import solve_ivp

def rhs(x, O):
    fprime = 4 * O[0] ** 3 * KO ** 4 / ((KO ** 4 + O[0] ** 4) ** 2)
    return (PO2 - O) / (1 + 4 * THb * fprime)


def twodrhs(x, dum):
    O, Y = dum
    X = THb - Y
    return [PO2 - O + 4*km1*Y - 4*k1*X*O**4, k1*X*O**4 - km1*Y]


# The figure in the text
sigmaO = 1.4  # Units of uMolar/mm Hg
KO = 30 * sigmaO
PO2 = 104 * sigmaO  # convert pressure to conc (units of uMolar)
THb = 2000  # 2 mM

# The figure in the text
L = 100
xspan = np.linspace(0, L, 2000)
init = [40 * sigmaO]
sol = solve_ivp(rhs, [0, L], init, t_eval=xspan)

plt.figure(1)
plt.plot(sol.t, sol.y[0] / sigmaO, 'r')  # plot the pressure, not the conc
plt.xlabel('dimensionless distance')
plt.ylabel(r'$[{\rm O}_2]$')


# This next part is for one of the Exercises.
# Find the solution without the equilibrium approximation. When solving
# this way, be careful with  initial conditions, and be sure 
# to start on the critical manifold. Otherwise, the solution takes some
# time to catch up and there is an offset in the solution. 

timescale = 10
k1 = timescale
km1 = (KO**4)*k1
Oinit = 40 * sigmaO
Yinit = THb * Oinit ** 4 / (KO**4 + Oinit**4)
init = [Oinit, Yinit]

L = 100
xspan = np.linspace(0, L, 200)
sol = solve_ivp(twodrhs, [0, L], init, method='Radau',t_eval=xspan)

plt.figure(1)
plt.plot(sol.t, sol.y[0] / sigmaO, '--b')  # plot the pressure, not the conc
plt.legend(['With equilibrium approximation', 'Without equilibrium approximation'])
plt.show()



